keps/0001-universal-operator.md

```diff
@@ -0,0 +1,28 @@
+# Maestro: A Universal Operator
+
+## Summary
+
+We propose a universal operator that generates and runs operators based on a standard data format. This universal operator manages the lifecycle of stateful workloads.
+
+## Motivation
+
+While Kubebuilder provides an abstraction around the plumbing for building operators, there is still a lot of work to implement operators in code. To combat this problem, Mesosphere built an SDK for DC/OS, as well as a package data format, that handles a lot of the common high level tasks in running stateful workloads and provides lifecycle hooks. The purpose of Maestro is to take this work and enable the same standard packaging format and lifecycle hooks to be applied to Kubernetes operators.
+
+## Goals
+
+* Make it easy to deploy stateful operators on Kubernetes using a standard data format
+* Enable re-use between DC/OS and Kubernetes SDK packages
+
+## Development
+
+Development will be located at [https://github.com/kubernetes-sigs/kubebuilder-maestro](https://github.com/kubernetes-sigs/kubebuilder-maestro).
+
+## Design
+
+The core of the Maestro Universal Operator is the "Framework" CRD. This CRD describes the implementation details of a framework, such as the deployment plan, components required, and other information needed to start a package. Initially, this Framework CRD's spec will be based on the [DC/OS Service SDK](https://mesosphere.github.io/dcos-commons/).
+
```

Comment:
It will be good to have a `hello-world` example to demonstrate the key components involved and the interaction between them. It will help if there is some description about how the data format description maps to the `k8s` primitives for stateful workloads will also make it easy to follow the proposal.

Reply:
> It will be good to have a `hello-world` example to demonstrate the key components involved and the interaction between them. It will help if there is some description about how the data format description maps to the `k8s` primitives for stateful workloads will also make it easy to follow the proposal.

I agree.  There at least two different ways to see this working out.

Reply:
Working on this with some more documentation. I'll ping you @deads2k when it's ready to review. A lot of this will be based on Mesosphere's existing SDK work here: https://github.com/mesosphere/dcos-commons. We're doing a little iteration on the spec to make it not specific to DC/OS and Mesos constructs.